crates/cext/src/sparse_observable.rs

```diff
@@ -0,0 +1,849 @@
+    unsafe { obs.indices_mut() }.as_mut_ptr()
+}
+
+/// @ingroup QkSparseObservable
+/// Get a pointer to the term boundaries.
+///
+/// This can be used to read and modify the observable's term boundaries. The resulting pointer is
+/// valid to read for ``qk_obs_num_terms(obs) + 1`` elements of size ``size_t``.
+///
+/// @param obs A pointer to the observable.
+///
+/// @return A pointer to the boundaries.
+///
+/// # Example
+///
+///     uint32_t num_qubits = 100;
+///     QkSparseObservable *obs = qk_obs_zero(num_qubits);
+///
+///     complex double coeff = 1;
+///     QkBitTerm bit_terms[3] = {QkBitTerm_X, QkBitTerm_Y, QkBitTerm_Z};
+///     uint32_t indices[3] = {0, 1, 2};
+///     QkSparseTerm term = {&coeff, 3, bit_terms, indices, num_qubits};
+///     qk_obs_add_term(obs, &term);
+///
+///     size_t num_terms = qk_obs_num_terms(obs);
+///     uint32_t *boundaries = qk_obs_boundaries(obs);
+///
+///     for (size_t i = 0; i < num_terms + 1; i++) {
+///         printf("boundary %i: %i\n", i, boundaries[i]);
+///     }
+///
+/// # Safety
+///
+/// Behavior is undefined ``obs`` is not a valid, non-null pointer to a ``QkSparseObservable``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_boundaries(obs: *mut SparseObservable) -> *mut usize {
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let obs = unsafe { mut_ptr_as_ref(obs) };
+
+    // SAFETY: Modifying the boundaries can leave the observable in an incoherent state. It is
+    // the responsibility of the user that the data is coherent.
+    unsafe { obs.boundaries_mut() }.as_mut_ptr()
+}
+
+/// @ingroup QkSparseObservable
+/// Get a pointer to the bit terms.
+///
+/// This can be used to read and modify the observable's bit terms. The resulting pointer is
+/// valid to read for ``qk_obs_len(obs)`` elements of size ``uint8_t``.
+///
+/// @param obs A pointer to the observable.
+///
+/// @return A pointer to the bit terms.
+///
+/// # Example
+///
+///     uint32_t num_qubits = 100;
+///     QkSparseObservable *obs = qk_obs_zero(num_qubits);
+///
+///     complex double coeff = 1;
+///     QkBitTerm bit_terms[3] = {QkBitTerm_X, QkBitTerm_Y, QkBitTerm_Z};
+///     uint32_t indices[3] = {0, 1, 2};
+///     QkSparseTerm term = {&coeff, 3, bit_terms, indices, num_qubits};
+///     qk_obs_add_term(obs, &term);
+///
+///     size_t len = qk_obs_len(obs);
+///     QkBitTerm *bits = qk_obs_bit_terms(obs);
+///
+///     for (size_t i = 0; i < len; i++) {
+///         printf("bit term %i: %i\n", i, bits[i]);
+///     }
+///
+/// # Safety
+///
+/// Behavior is undefined ``obs`` is not a valid, non-null pointer to a ``QkSparseObservable``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_bit_terms(obs: *mut SparseObservable) -> *mut BitTerm {
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let obs = unsafe { mut_ptr_as_ref(obs) };
+
+    obs.bit_terms_mut().as_mut_ptr()
+}
```

Comment:
We need to be super clear in here that you _must not_ write invalid `BitTerm` values into the pointer - it's pretty immediate undefined behaviour.

Reply:
Extended the `Safety` section on this in 624c3dd